File: python/t402/src/t402/bridge/scan.py
"""LayerZero Scan API Client for tracking cross-chain messages."""

import asyncio
from typing import Optional

import httpx

from .constants import (
    DEFAULT_POLL_INTERVAL,
    DEFAULT_TIMEOUT,
    LAYERZERO_SCAN_BASE_URL,
)
from .types import (
    LayerZeroMessage,
    LayerZeroMessageStatus,
    WaitForDeliveryOptions,
)


class LayerZeroScanClient:
    """LayerZero Scan API Client.

    Provides tracking for cross-chain messages via LayerZero Scan.

    Example:
        ```python
        from t402.bridge import LayerZeroScanClient

        client = LayerZeroScanClient()

        # Get message status
        message = await client.get_message(message_guid)
        print(f"Status: {message.status}")

        # Wait for delivery
        delivered = await client.wait_for_delivery(
            message_guid,
            on_status_change=lambda s: print(f"Status: {s}")
        )
        print(f"Delivered! Dest TX: {delivered.dst_tx_hash}")
        ```
    """

    def __init__(self, base_url: str = LAYERZERO_SCAN_BASE_URL) -> None:
        """Create a new LayerZero Scan client.

        Args:
            base_url: API base URL (default: production endpoint)
        """
        self.base_url = base_url
        self._client: Optional[httpx.AsyncClient] = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client."""
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=30.0)
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    async def get_message(self, guid: str) -> LayerZeroMessage:
        """Get message by GUID.

        Args:
            guid: LayerZero message GUID

        Returns:
            Message details including status

        Raises:
            ValueError: If message not found
            httpx.HTTPError: If API error
        """
        client = await self._get_client()
        url = f"{self.base_url}/messages/guid/{guid}"

        response = await client.get(url, headers={"Accept": "application/json"})

        if response.status_code == 404:
            raise ValueError(f"Message not found: {guid}")

        response.raise_for_status()
        data = response.json()
        return self._map_api_response(data)

    async def get_messages_by_wallet(
        self, address: str, limit: int = 20
    ) -> list[LayerZeroMessage]:
        """Get messages by wallet address.

        Args:
            address: Wallet address that initiated messages
            limit: Maximum number of messages to return (default: 20)

        Returns:
            List of messages

        Raises:
            httpx.HTTPError: If API error
        """
        client = await self._get_client()
        url = f"{self.base_url}/messages/wallet/{address}?limit={limit}"

        response = await client.get(url, headers={"Accept": "application/json"})
        response.raise_for_status()

        data = response.json()
        messages_data = data.get("messages") or data.get("data") or []
        return [self._map_api_response(msg) for msg in messages_data]

    async def wait_for_delivery(
        self,
        guid: str,
        options: Optional[WaitForDeliveryOptions] = None,
    ) -> LayerZeroMessage:
        """Poll message status until delivered or failed.

        Args:
            guid: LayerZero message GUID
            options: Wait configuration options

        Returns:
            Final message state (DELIVERED)

        Raises:
            ValueError: If message fails, is blocked, or times out
        """
        timeout_ms = DEFAULT_TIMEOUT
        poll_interval_ms = DEFAULT_POLL_INTERVAL
        on_status_change = None

        if options is not None:
            if options.timeout > 0:
                timeout_ms = options.timeout
            if options.poll_interval > 0:
                poll_interval_ms = options.poll_interval
            on_status_change = options.on_status_change

        timeout_sec = timeout_ms / 1000
        poll_interval_sec = poll_interval_ms / 1000
        elapsed = 0.0
        last_status: Optional[LayerZeroMessageStatus] = None

        while elapsed < timeout_sec:
            try:
                message = await self.get_message(guid)

                # Notify on status change
                if message.status != last_status:
                    last_status = message.status
                    if on_status_change is not None:
                        on_status_change(message.status)

                # Check terminal states
                if message.status == LayerZeroMessageStatus.DELIVERED:
                    return message

                if message.status == LayerZeroMessageStatus.FAILED:
                    raise ValueError(f"Bridge message failed: {guid}")

                if message.status == LayerZeroMessageStatus.BLOCKED:
                    raise ValueError(f"Bridge message blocked by DVN: {guid}")

                # Continue polling for INFLIGHT/CONFIRMING
                await asyncio.sleep(poll_interval_sec)
                elapsed += poll_interval_sec

            except ValueError as e:
                # Message not yet indexed, retry
                if "not found" in str(e).lower():
                    await asyncio.sleep(poll_interval_sec)
                    elapsed += poll_interval_sec
                    continue
                raise

        raise ValueError(f"Timeout waiting for message delivery: {guid}")

    async def is_delivered(self, guid: str) -> bool:
        """Check if a message has been delivered.

        Args:
            guid: LayerZero message GUID

        Returns:
            True if delivered, False otherwise
        """
        try:
            message = await self.get_message(guid)
            return message.status == LayerZeroMessageStatus.DELIVERED
        except ValueError:
            return False

    def _map_api_response(self, data: dict) -> LayerZeroMessage:
        """Map API response to LayerZeroMessage."""
        return LayerZeroMessage(
            guid=data.get("guid") or data.get("messageGuid") or "",
            src_eid=int(data.get("srcEid") or data.get("srcChainId") or 0),
            dst_eid=int(data.get("dstEid") or data.get("dstChainId") or 0),
            src_ua_address=data.get("srcUaAddress") or data.get("srcAddress") or "",
            dst_ua_address=data.get("dstUaAddress") or data.get("dstAddress") or "",
            src_tx_hash=data.get("srcTxHash") or "",
            dst_tx_hash=data.get("dstTxHash"),
            status=LayerZeroMessageStatus(data.get("status") or "INFLIGHT"),
            src_block_number=int(data.get("srcBlockNumber") or 0),
            dst_block_number=int(data.get("dstBlockNumber")) if data.get("dstBlockNumber") else None,
            created=data.get("created") or data.get("createdAt") or "",
            updated=data.get("updated") or data.get("updatedAt") or "",
        )


def create_layerzero_scan_client(
    base_url: str = LAYERZERO_SCAN_BASE_URL,
) -> LayerZeroScanClient:
    """Create a LayerZero Scan client.

    Args:
        base_url: Optional custom API base URL

    Returns:
        New client instance
    """
    return LayerZeroScanClient(base_url)

File: typescript/packages/mechanisms/evm/src/bridge/scan.ts
/**
 * LayerZero Scan API Client
 *
 * Provides tracking for cross-chain messages via LayerZero Scan.
 *
 * @see https://docs.layerzero.network/v2/tools/api/scan/mainnet
 *
 * @example
 * ```typescript
 * import { LayerZeroScanClient } from '@t402/evm';
 *
 * const scanClient = new LayerZeroScanClient();
 *
 * // Get message status
 * const message = await scanClient.getMessage(messageGuid);
 * console.log(`Status: ${message.status}`);
 *
 * // Wait for delivery
 * const delivered = await scanClient.waitForDelivery(messageGuid, {
 *   onStatusChange: (status) => console.log(`Status changed: ${status}`),
 * });
 * console.log(`Delivered! Destination TX: ${delivered.dstTxHash}`);
 * ```
 */

import type {
  LayerZeroMessage,
  LayerZeroMessageStatus,
  WaitForDeliveryOptions,
} from "./types.js";

/**
 * LayerZero Scan API base URL
 */
export const LAYERZERO_SCAN_BASE_URL = "https://scan.layerzero-api.com/v1";

/**
 * Default timeout for waiting (10 minutes)
 */
const DEFAULT_TIMEOUT = 600_000;

/**
 * Default poll interval (10 seconds)
 */
const DEFAULT_POLL_INTERVAL = 10_000;

/**
 * LayerZero Scan API Client
 *
 * Use this client to track the status of cross-chain messages sent via LayerZero.
 */
export class LayerZeroScanClient {
  private readonly baseUrl: string;

  /**
   * Create a new LayerZero Scan client
   *
   * @param baseUrl - API base URL (default: production endpoint)
   */
  constructor(baseUrl: string = LAYERZERO_SCAN_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  /**
   * Get message by GUID
   *
   * @param guid - LayerZero message GUID
   * @returns Message details including status
   * @throws Error if message not found or API error
   */
  async getMessage(guid: string): Promise<LayerZeroMessage> {
    const url = `${this.baseUrl}/messages/guid/${guid}`;

    const response = await fetch(url, {
      method: "GET",
      headers: {
        Accept: "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw new Error(`Message not found: ${guid}`);
      }
      throw new Error(`LayerZero Scan API error: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    return this.mapApiResponse(data);
  }

  /**
   * Get messages by wallet address
   *
   * @param address - Wallet address that initiated messages
   * @param limit - Maximum number of messages to return (default: 20)
   * @returns Array of messages
   */
  async getMessagesByWallet(
    address: string,
    limit: number = 20,
  ): Promise<LayerZeroMessage[]> {
    const url = `${this.baseUrl}/messages/wallet/${address}?limit=${limit}`;

    const response = await fetch(url, {
      method: "GET",
      headers: {
        Accept: "application/json",
      },
    });

    if (!response.ok) {
      throw new Error(`LayerZero Scan API error: ${response.status} ${response.statusText}`);
    }

    const data = (await response.json()) as Record<string, unknown>;
    const messages = (data.messages ?? data.data ?? []) as unknown[];
    return messages.map((msg: unknown) => this.mapApiResponse(msg));
  }

  /**
   * Poll message status until delivered or failed
   *
   * @param guid - LayerZero message GUID
   * @param options - Wait configuration options
   * @returns Final message state (DELIVERED)
   * @throws Error if message fails, is blocked, or times out
   */
  async waitForDelivery(
    guid: string,
    options: WaitForDeliveryOptions = {},
  ): Promise<LayerZeroMessage> {
    const {
      timeout = DEFAULT_TIMEOUT,
      pollInterval = DEFAULT_POLL_INTERVAL,
      onStatusChange,
    } = options;

    const startTime = Date.now();
    let lastStatus: LayerZeroMessageStatus | null = null;

    while (Date.now() - startTime < timeout) {
      try {
        const message = await this.getMessage(guid);

        // Notify on status change
        if (lastStatus !== message.status) {
          lastStatus = message.status;
          onStatusChange?.(message.status);
        }

        // Check terminal states
        if (message.status === "DELIVERED") {
          return message;
        }

        if (message.status === "FAILED") {
          throw new Error(`Bridge message failed: ${guid}`);
        }

        if (message.status === "BLOCKED") {
          throw new Error(`Bridge message blocked by DVN: ${guid}`);
        }

        // Continue polling for INFLIGHT/CONFIRMING
        await this.sleep(pollInterval);
      } catch (error) {
        // Message not yet indexed, retry
        if ((error as Error).message.includes("not found")) {
          await this.sleep(pollInterval);
          continue;
        }
        throw error;
      }
    }

    throw new Error(`Timeout waiting for message delivery: ${guid}`);
  }

  /**
   * Check if a message has been delivered
   *
   * @param guid - LayerZero message GUID
   * @returns True if delivered, false otherwise
   */
  async isDelivered(guid: string): Promise<boolean> {
    try {
      const message = await this.getMessage(guid);
      return message.status === "DELIVERED";
    } catch {
      return false;
    }
  }

  /**
   * Map API response to our interface
   */
  private mapApiResponse(data: unknown): LayerZeroMessage {
    const msg = data as Record<string, unknown>;
    return {
      guid: (msg.guid ?? msg.messageGuid ?? "") as string,
      srcEid: (msg.srcEid ?? msg.srcChainId ?? 0) as number,
      dstEid: (msg.dstEid ?? msg.dstChainId ?? 0) as number,
      srcUaAddress: (msg.srcUaAddress ?? msg.srcAddress ?? "") as string,
      dstUaAddress: (msg.dstUaAddress ?? msg.dstAddress ?? "") as string,
      srcTxHash: (msg.srcTxHash ?? "") as string,
      dstTxHash: (msg.dstTxHash ?? undefined) as string | undefined,
      status: (msg.status ?? "INFLIGHT") as LayerZeroMessageStatus,
      srcBlockNumber: (msg.srcBlockNumber ?? 0) as number,
      dstBlockNumber: (msg.dstBlockNumber ?? undefined) as number | undefined,
      created: (msg.created ?? msg.createdAt ?? "") as string,
      updated: (msg.updated ?? msg.updatedAt ?? "") as string,
    };
  }

  /**
   * Sleep helper
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Create a LayerZero Scan client
 *
 * @param baseUrl - Optional custom API base URL
 */
export function createLayerZeroScanClient(
  baseUrl?: string,
): LayerZeroScanClient {
  return new LayerZeroScanClient(baseUrl);
}
